src/autodiff.jl

```diff
@@ -66,86 +66,127 @@ struct StateJacobian{VI, VT, MT, SMT, VP, VD, SubT, SubD} <: AbstractJacobian
-
-        t1sseeds = A{ForwardDiff.Partials{ncolor,Float64}}(undef, nmap)
-        _init_seed!(t1sseeds, coloring, ncolor, nmap)
-
-        compressedJ = MT(zeros(Float64, ncolor, nmap))
-        # Views
-        varx = view(x, map)
-        t1svarx = view(t1sx, map)
-        VP = typeof(t1sseeds)
-        VD = typeof(t1sx)
-        return new{VI, VT, MT, SMT, VP, VD, typeof(varx), typeof(t1svarx)}(
-            J, compressedJ, coloring, t1sseeds, t1sF, x, t1sx, map, varx, t1svarx
-        )
+    mappv = [i + nv_m for i in pv]
+    mappq = [i + nv_m for i in pq]
+    # Ordering for x is (θ_pv, θ_pq, v_pq)
+    map = VI(vcat(mappv, mappq, pq))
+    nmap = size(map,1)
+
+    # Used for sparsity detection with randomized inputs
+    function residual_jacobian(V, Ybus, pv, pq)
+        n = size(V, 1)
+        Ibus = Ybus*V
+        diagV       = sparse(1:n, 1:n, V, n, n)
+        diagIbus    = sparse(1:n, 1:n, Ibus, n, n)
+        diagVnorm   = sparse(1:n, 1:n, V./abs.(V), n, n)
+
+        dSbus_dVm = diagV * conj(Ybus * diagVnorm) + conj(diagIbus) * diagVnorm
+        dSbus_dVa = 1im * diagV * conj(diagIbus - Ybus * diagV)
+
+        j11 = real(dSbus_dVa[[pv; pq], [pv; pq]])
+        j12 = real(dSbus_dVm[[pv; pq], pq])
+        j21 = imag(dSbus_dVa[pq, [pv; pq]])
+        j22 = imag(dSbus_dVm[pq, pq])
+
+        J = [j11 j12; j21 j22]
+    end
+
+    # Need a host arrays for the sparsity detection below
+    spmap = Vector(map)
+    hybus_re = Spmat{Vector{Int}, Vector{Float64}}(ybus_re)
+    hybus_im = Spmat{Vector{Int}, Vector{Float64}}(ybus_im)
+    n = nv_a
+    Yre = SparseMatrixCSC{Float64,Int64}(n, n, hybus_re.colptr, hybus_re.rowval, hybus_re.nzval)
+    Yim = SparseMatrixCSC{Float64,Int64}(n, n, hybus_im.colptr, hybus_im.rowval, hybus_im.nzval)
+    Y = Yre .+ 1im .* Yim
+    # Randomized inputs
+    Vre = Float64.([i for i in 1:n])
+    Vim = Float64.([i for i in n+1:2*n])
+    V = Vre .+ 1im .* Vim
+    J = residual_jacobian(V, Y, pv, pq)
+    coloring = VI(matrix_colors(J))
+    ncolor = size(unique(coloring),1)
+    if F isa CuArray
+        J = CuSparseMatrixCSR(J)
+    end
+    t1s{N} = ForwardDiff.Dual{Nothing,Float64, N} where N
+    x = VT(zeros(Float64, nv_m + nv_a))
+    t1sx = A{t1s{ncolor}}(x)
+    t1sF = A{t1s{ncolor}}(zeros(Float64, nmap))
+
+    t1sseeds = A{ForwardDiff.Partials{ncolor,Float64}}(undef, nmap)
+    _init_seed!(t1sseeds, coloring, ncolor, nmap)
+
+    compressedJ = MT(zeros(Float64, ncolor, nmap))
+    # Views
+    varx = view(x, map)
+    t1svarx = view(t1sx, map)
+    VP = typeof(t1sseeds)
+    VD = typeof(t1sx)
+    return StateJacobian{VI, VT, MT, SMT, VP, VD, typeof(varx), typeof(t1svarx)}(
+        J, compressedJ, coloring, t1sseeds, t1sF, x, t1sx, map, varx, t1svarx
+    )
+end
+
+function StateJacobian(jac::StateJacobian, device)
+    ncolor = size(unique(jac.coloring),1)
+    t1s{N} = ForwardDiff.Dual{Nothing,Float64, N} where N
+    if device == CUDADevice()
```

Comment:
Again, not related to this PR, but I think in the long term we should find a way to automatically dispatch the proper Julia types depending on the device where we are deporting the computation upon.

Reply:
Yes, it's the CSR CSC issue. we'd have to introduce an `AbstractSparseMatrix` type. We could also here handle it such that we instantiate it for every device type available.